### PR TITLE
accept require.resolve implementation as option to importLocation et al.

### PR DESCRIPTION
**Pull request by naugtur**
Receiving the implementation externally makes more sense than supporting require.resolve functionality endoEndo.

This PR accepts the implementation and, if not provided, defaults to a similar error Node would throw.

### CODE REVIEW

File: packages/compartment-mapper/src/parse-cjs-shared-export-wrapper.js [hunk 1]
```diff
@@ -131,7 +140,20 @@ export const wrap = (moduleEnvironmentRecord, compartment, resolvedImports) => {
...
+  };
+  if (requireResolve) {
+    require.resolve = freeze((specifier, options) =>
+      requireResolve(location, specifier, options),
+    );
+  } else {
+    require.resolve = freeze(specifier => {
+      const error = Error(`Cannot find module '${specifier}'`);
+      defineProperty(error, 'code', { value: 'MODULE_NOT_FOUND' });
+      throw error;
```
Comment (naugtur) [inline packages/compartment-mapper/src/parse-cjs-shared-export-wrapper.js:152]: Question: Should this be indistinguishable from an error Node would throw, or rather match it loosely but leave a hint somewhere to it resulting from a lack of requireResolve option being passed to Endo?

File: packages/compartment-mapper/test/scaffold.js [hunk 2]
```diff
@@ -87,6 +88,8 @@ export function scaffold(
 
     const application = await loadLocation(readPowers, fixture, {
       dev: true,
+      requireResolve: (from, specifier, options) =>
+        createRequire(from).resolve(specifier, options),
```
Comment (naugtur) [inline packages/compartment-mapper/test/scaffold.js:92]: I like how easy it is to provide the implementation.
A more secure implementation would use `options.paths` to limit the search space and guard return values by checking with `String.startsWith`that it's not reaching beyond the project folder.

A cache lookup of known needed paths for specifiers is also an option.

File: packages/compartment-mapper/test/test-cjs-compat.js [hunk 3]
```diff
@@ -26,10 +26,35 @@ const assertFixture = (t, { namespace }) => {
...
+      '/skipped/fixtures-cjs-compat/node_modules/require-resolve/nested/index.js',
+      '/skipped/fixtures-cjs-compat/node_modules/require-resolve/nested/file.js',
+      '/skipped/fixtures-cjs-compat/node_modules/require-resolve/nested/file.js.map',
+      "Cannot find module './nested/file.missing'",
+      '/skipped/fixtures-cjs-compat/node_modules/app/index.js',
+      'fs',
+      '/skipped/fixtures-cjs-compat/node_modules/nested-export/callBound.js',
+    ]);
+  } else {
+    t.deepEqual(results.requireResolvePaths, [
```
Comment (naugtur) [inline packages/compartment-mapper/test/test-cjs-compat.js:42]: This is what it looks like when requireResolve implementation is not provided.

File: packages/compartment-mapper/src/parse-cjs-shared-export-wrapper.js (hunk 1 above)
Comment (naugtur) [reply packages/compartment-mapper/src/parse-cjs-shared-export-wrapper.js:152]: from a meeting: add the second line to the error mesage explaining requireResolve needs to be passed

**Review by kriskowal — COMMENTED** [review]
I would like to fold requireResolve into the optional properties of readPowers. Otherwise, this looks good to me.

File: packages/compartment-mapper/src/parse-cjs-shared-export-wrapper.js [hunk 4]
```diff
@@ -131,7 +140,22 @@ export const wrap = (moduleEnvironmentRecord, compartment, resolvedImports) => {
     } else {
       return namespace;
     }
-  });
+  };
+  if (readPowers && readPowers.requireResolve) {
+    require.resolve = freeze((specifier, options) =>
+      readPowers.requireResolve(location, specifier, options),
```
Comment (naugtur) [inline packages/compartment-mapper/src/parse-cjs-shared-export-wrapper.js:146]: Now all that's left to do is convince Typescript I can call this function if I check the field exists even thought it's optional on readPowers

**Comment by mhofman** [comment]
I'm not sure I have much context to review this PR. Is there anything specific you'd like me to take a look at?

File: packages/compartment-mapper/src/parse-cjs-shared-export-wrapper.js (hunk 4 above)
Comment (naugtur) [reply packages/compartment-mapper/src/parse-cjs-shared-export-wrapper.js:146]: it stopped complaining when I did this
```js
 if (typeof readPowers === 'object' && readPowers.requireResolve) {
    const { requireResolve } = readPowers;
    require.resolve = freeze((specifier, options) =>
      requireResolve(location, specifier, options),
    );
```
I don't appreciate what TS is trying to do for me here

**Comment by naugtur** [comment]
@kriskowal This is ready IMHO, I need *the* final review.

**Comment by mhofman** [comment]
> @mhofman Automatons said you had overlap with the codebase that this touches

Automatons are mistaken, I'll remove myself as reviewer.

File: packages/compartment-mapper/src/import-hook.js [hunk 5]
```diff
@@ -176,7 +176,9 @@ export const makeImportHookMaker = (
             candidateSpecifier,
             moduleLocation,
             packageLocation,
-            readPowers,
+            {
+              readPowers,
+            },
```
Comment (kriskowal) [inline packages/compartment-mapper/src/import-hook.js:181]: Is this envelope still necessary?

File: packages/compartment-mapper/src/parse-cjs-shared-export-wrapper.js [hunk 6]
```diff
@@ -131,7 +140,23 @@ export const wrap = (moduleEnvironmentRecord, compartment, resolvedImports) => {
...
+  };
+  if (typeof readPowers === 'object' && readPowers.requireResolve) {
+    const { requireResolve } = readPowers;
+    require.resolve = freeze((specifier, options) =>
+      requireResolve(location, specifier, options),
+    );
+  } else {
+    require.resolve = freeze(specifier => {
+      const error = Error(
+        `Cannot find module '${specifier}'\nPass ReadPowers with a requireResolve function to provide require.resolve`,
```
Comment (kriskowal) [inline packages/compartment-mapper/src/parse-cjs-shared-export-wrapper.js:152]: Maybe `Add requireResolve to Endo Compartment Mapper readPowers.`

**Review by kriskowal — APPROVED** [review]

**Pull request merged**